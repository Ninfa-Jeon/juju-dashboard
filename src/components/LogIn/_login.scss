@import "../../scss/functions/z-index";

.login {
  background-color: rgba(255, 255, 255, 0.75);
  height: 100vh;
  left: 0;
  position: fixed;
  top: 0;
  width: 100vw;
  z-index: z("lambda");

  &__inner {
    display: grid;
    left: 50%;
    margin: auto;
    position: fixed;
    top: 50%;
    transform: translate(-50%, -50%);
    width: 280px;

    .p-button--positive,
    .p-button--neutral {
      margin: 0 auto;
      width: 100%;
    }

    .p-icon--spinner {
      top: -2px;
    }
  }

  &__logo {
    display: block;
    margin: 0 auto 1rem;
    width: 110px;
  }
}
