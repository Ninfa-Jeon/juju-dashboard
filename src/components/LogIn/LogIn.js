import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector, useStore } from "react-redux";
import { isLoggedIn, getBakery, getConfig, getLoginError } from "app/selectors";
import { connectAndStartPolling, storeUserPass } from "app/actions";

import Spinner from "@canonical/react-components/dist/components/Spinner";

import logo from "static/images/logo/logo-black-on-white.svg";

import "./_login.scss";

export default function LogIn({ children }) {
  const { identityProviderAvailable } = useSelector(getConfig);
  const userIsLoggedIn = useSelector(isLoggedIn);
  const loginError = useSelector(getLoginError);

  if (!userIsLoggedIn) {
    return (
      <>
        <div className="login">
          <div className="login__inner p-card--highlighted">
            <img className="login__logo" src={logo} alt="JAAS logo" />
            {identityProviderAvailable ? (
              <IdentityProviderForm />
            ) : (
              <UserPassForm />
            )}
            {generateErrorMessage(loginError)}
          </div>
        </div>
        <main>{children}</main>
      </>
    );
  }
  return children;
}

/**
  Generates the necessary elements to render the error message.
  @param {String} loginError The error message from the store.
  @returns {Object} A component for the error message.
*/
function generateErrorMessage(loginError) {
  if (!loginError) {
    return null;
  }
  let loginErrorMessage = null;
  switch (loginError) {
    case '"user-" is not a valid user tag':
      loginErrorMessage = "Invalid user name";
      break;
    case "invalid entity name or password":
      loginErrorMessage = "invalid user name or password";
      break;
    default:
      loginErrorMessage = loginError;
  }
  return (
    <p className="error-message">
      <i className="p-icon--error" />
      {loginErrorMessage}
    </p>
  );
}

function IdentityProviderForm() {
  const userIsLoggedIn = useSelector(isLoggedIn);

  const visitURL = useSelector((state) => {
    if (!userIsLoggedIn) {
      const root = state.root;
      if (root && root.visitURL) {
        return root.visitURL;
      }
    }
  });

  return <Button visitURL={visitURL}></Button>;
}

function UserPassForm() {
  const dispatch = useDispatch();
  const store = useStore();
  const bakery = useSelector(getBakery);
  const focus = useRef();

  function handleSubmit(e) {
    e.preventDefault();
    const elements = e.currentTarget.elements;
    const user = elements.username.value;
    const password = elements.password.value;
    dispatch(storeUserPass({ user, password }));
    dispatch(connectAndStartPolling(store, bakery));
  }

  useEffect(() => {
    focus.current.focus();
  }, []);

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="username">Username</label>
      <input type="text" name="username" id="username" ref={focus} />
      <label htmlFor="password">Password</label>
      <input type="password" name="password" id="password" />
      <button className="p-button--positive" type="submit">
        Log in to the dashboard
      </button>
    </form>
  );
}

function Button({ visitURL }) {
  if (visitURL) {
    return (
      <a
        className="p-button--positive"
        href={visitURL}
        rel="noopener noreferrer"
        target="_blank"
      >
        Log in to the dashboard
      </a>
    );
  } else {
    return (
      <button className="p-button--neutral" disabled>
        <Spinner text="Connecting..." />
      </button>
    );
  }
}
