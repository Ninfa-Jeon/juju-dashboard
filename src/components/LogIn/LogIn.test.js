import React from "react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import dataDump from "testing/complete-redux-store-dump";

import LogIn from "./LogIn";

const mockStore = configureStore([]);

describe("LogIn", () => {
  it("renders a 'connecting' message while connecting", () => {
    const store = mockStore({
      root: {}
    });
    const wrapper = mount(
      <Provider store={store}>
        <LogIn>App content</LogIn>
      </Provider>
    );
    expect(wrapper.find(".p-button").text()).toBe("Connecting...");
    expect(wrapper.find("LogIn main").text()).toBe("App content");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a login UI if the user is not logged in", () => {
    const store = mockStore({
      root: {
        visitURL: "I am a url"
      }
    });
    const wrapper = mount(
      <Provider store={store}>
        <LogIn>App content</LogIn>
      </Provider>
    );
    expect(wrapper.find(".p-button--positive").text()).toBe(
      "Log in to view your models"
    );
    expect(wrapper.find("LogIn main").text()).toBe("App content");
    expect(wrapper).toMatchSnapshot();
  });
});
