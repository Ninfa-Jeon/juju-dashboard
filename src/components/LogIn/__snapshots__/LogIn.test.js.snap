// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogIn renders a login UI if the user is not logged in 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LogIn>
    <div
      className="login"
    >
      <div
        className="login__inner"
      >
        <img
          alt="JAAS logo"
          className="login__logo"
          src="logo-black-on-white.svg"
        />
        <a
          className="p-button--positive"
          rel="noopener noreferrer"
          target="_blank"
        >
          Log in to view your models
        </a>
      </div>
    </div>
  </LogIn>
</Provider>
`;

exports[`LogIn renders it's children if the user is logged in 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LogIn>
    Hello world!
  </LogIn>
</Provider>
`;
